mdsuite/graph_modules/molecular_graph.py

```diff
@@ -247,14 +256,224 @@ def check_a_in_b(a, b):
+
+        return particle_groups
+
+
+def build_smiles_graph(smiles_string: str) -> tuple:
+    """
+    Build molecular graphs from SMILES strings.
+
+    Parameters
+    ----------
+    smiles_string : str
+            SMILES string to use in the graph construction.
+
+    Returns
+    -------
+    smiles_graph :
+            Graph object returned by PySmiles
+    species : dict
+            A dict object containing species information about the molecule.
+    """
+    mol = read_smiles(smiles_string, explicit_hydrogen=True)
+    data = mol.nodes
+    species = {}
+    for i in range(len(data)):
+        item = data[i].get("element")
+        if item in species:
+            species[item] += 1
+        else:
+            species[item] = 0
+            species[item] += 1
+
+    return mol, species
+
+
+def _apply_system_cutoff(input_tensor: tf.Tensor, cutoff: float) -> tf.Tensor:
+    """
+    Enforce a cutoff on a tensor.
+
+    In this context the cutoff is used to identify bonded atoms. We argue
+    that the closest atoms will be bonded as defined by a cutoff.
+    Constructing the mask of closest atoms will allow for the bonded ones to be
+    identified.
+
+    Parameters
+    ----------
+    input_tensor : tf.Tensor
+            Tensor of any size or shape to be masked. In our case it is a distance
+            tensor of the atoms in a configuration.
+    cutoff : float
+            Cutoff to use in the mask. If a distance is greater than this cutoff it
+            is marked as 0, if not, it is 1.
+
+    Returns
+    -------
+    masked_tensor : tf.Tensor
+            A tensor of ones and zeros where 1s corresponded to 'bonded' particles
+            and 0s indicated no bonding. Note, the diagonals of this tensor are
+            set to 0 as a particle cannot bond itself.
+    """
+
+    cutoff_mask = tf.cast(
+        tf.less(input_tensor, cutoff), dtype=tf.int16
+    )  # Construct the mask
+
+    return tf.linalg.set_diag(cutoff_mask, np.zeros(len(input_tensor)))
+
+
+def get_neighbour_list(positions: tf.Tensor, cell: list = None) -> tf.Tensor:
+    """
+    Generate the neighbour list
+
+    Parameters
+    ----------
+    positions: tf.Tensor
+        Tensor with shape (number_of_configurations, n_atoms, 3)
+        representing the coordinates
+    cell: list
+        If periodic boundary conditions are used, please supply the cell
+        dimensions, e.g. [13.97, 13.97, 13.97]. If the cell is provided
+        minimum image convention will be applied!
+
+    Returns
+    -------
+    neighbour_list : tf.Tensor
+            Neighbour list for a single configuration.
+
+    """
+    r_ij_matrix = tf.reshape(positions, (1, len(positions), 3)) - tf.reshape(
+        positions, (len(positions), 1, 3)
+    )
+
+    # Pretty sure we never need min image for mapping.
+    if cell:
+        r_ij_matrix -= tf.math.rint(r_ij_matrix / cell) * cell
+    return tf.norm(r_ij_matrix, ord="euclidean", axis=2)
```

Comment:
We should put this method in one place in utils and use it in ADF / RDF / MolecularGraph and avoid making copies of it.

Reply:
I agree, for this one it was easier at the start to not have the upper trianlge only due to the decomposition and debugging but in the future that is definitely advantageous.

Reply:
if there is a chance now to reduce code duplication, we should not postpone (and forget about) it

Reply:
The issue is that the two methods differ slightly and it is not a small effort to change it, I can make an issue though. Furthermore, some of the issues solved in this PR are extremely critical so if someone else wants to make this change they can now but otherwise I think it should wait for a separate approach where all the relevant methods are checked and tested.